fe/fe-core/src/main/java/com/starrocks/journal/JournalWriter.java

```diff
@@ -0,0 +1,218 @@
+                    String msg = "got interrupted exception when trying to write one batch, will exit now.";
+                    LOG.error(msg, e);
+                    // TODO we should exit gracefully on InterruptedException
+                    Util.stdoutWithTime(msg);
+                    System.exit(-1);
+                }
+            }
+        };
+        d.start();
+    }
+
+    protected void writeOneBatch() throws InterruptedException {
+        // waiting if necessary until an element becomes available
+        currentJournal = journalQueue.take();
+        long nextJournalId = nextVisibleJournalId;
+        initBatch();
+
+        try {
+            this.journal.batchWriteBegin();
+
+            while (true) {
+                journal.batchWriteAppend(nextJournalId, currentJournal.getBuffer());
+                currentBatchTasks.add(currentJournal);
+                nextJournalId += 1;
+
+                if (shouldCommitNow()) {
+                    break;
+                }
+
+                currentJournal = journalQueue.take();
+            }
+        } catch (JournalException e) {
+            // abort current task
+            LOG.warn("failed to write batch, will abort current journal {} and commit", currentJournal, e);
+            abortJournalTask(currentJournal, e.getMessage());
+        } finally {
+            try {
+                // commit
+                journal.batchWriteCommit();
+                LOG.debug("batch write commit success, from {} - {}", nextVisibleJournalId, nextJournalId);
+                nextVisibleJournalId = nextJournalId;
+                markCurrentBatchSucceed();
+            } catch (JournalException e) {
+                // abort
+                LOG.warn("failed to commit batch, will abort current {} journals.",
+                        currentBatchTasks.size(), e);
+                try {
+                    journal.batchWriteAbort();
+                } catch (JournalException e2) {
+                    LOG.warn("failed to abort batch, will ignore and continue.", e);
+                }
+                abortCurrentBatch(e.getMessage());
+            }
+        }
+
+        rollJournalAfterBatch();
+
+        updateBatchMetrics();
+    }
+
+    private void initBatch() {
+        startTimeNano = System.nanoTime();
+        uncommittedEstimatedBytes = 0;
+        currentBatchTasks.clear();
+    }
+
+    private void markCurrentBatchSucceed() {
+        for (JournalTask t : currentBatchTasks) {
+            t.markSucceed();
+        }
+    }
+
+    private void abortCurrentBatch(String errMsg) {
+        for (JournalTask t : currentBatchTasks) {
+            abortJournalTask(t, errMsg);
+        }
+    }
+
+    /**
+     * We should notify the caller to rollback or report error on abort, like this.
+     *
+     * task.markAbort();
+     *
+     * But now we have to exit for historical reason.
+     * Note that if we exit here, the finally clause(commit current batch) will not be executed.
```

Comment:
what dose this comment mean? if we do not exit here, will the finally clause be executed again?

Reply:
If we do not exit here, the finally clause will be executed, current batch will be committed.
But since we exit here, the finally clause will not be executed, current batch will not be committed.

Reply:
here `exit` means to exit the process by calling `System.exit()`.